vine: `dyn fn`s

One issue with Rust's closures is that they capture their variables immediately when they are defined, rather than when they are used. For example, this code is invalid:
```rs
let mut x = 2;
let get_x_squared = || x * x;
get_x_squared();
x += 1;
get_x_squared();
```
```
error[E0506]: cannot assign to `x` because it is borrowed
 --> src/main.rs:5:1
  |
3 | let get_x_squared = || x * x;
  |                     -- - borrow occurs due to use in closure
  |                     |
  |                     `x` is borrowed here
4 | get_x_squared();
5 | x += 1;
  | ^^^^^^ `x` is assigned to here but it was already borrowed
6 | get_x_squared();
  | ------------- borrow later used here
```
To get it to work properly, you would need to redefine `get_x_squared` like so:
```rs
let mut x = 2;
let get_x_squared = || x * x;
get_x_squared();
x += 1;
let get_x_squared = || x * x;
get_x_squared();
```

But this rather defeats the point of defining the closure in the first place.

Vine does not yet have proper closures (function expressions can capture variables, but if they mutate any variables, they must be used linearly; i.e. cannot be called multiple times).

However, when vine does get proper closure support, it would have a similar issue.

One (ugly) solution in Rust is to use a macro:
```rs
let mut x = 2;
macro_rules! get_x_squared { () => { x * x } }
get_x_squared!(); // 4
x += 1;
get_x_squared!(); // 9
```

This has a number of downsides; the code needs to get compiled several times, scoping works strangely, etc.

In Vine, this issue could be solved with `dyn fn`s:

```rs
let x = 2;
dyn fn get_x_squared() { x * x }
get_x_squared(); // 4
x += 1;
get_x_squared(); // 9
```

`dyn fn`s can capture and mutate variables, but they do this at each place they're used, rather than once upfront. This allows the above code block to function correctly.

You could also have multiple `dyn fn` that reference the same variables:
```rs
let x = 2;
dyn fn get_x_squared() { x * x }
dyn fn increment_x() { x += 1 }
get_x_squared(); // 4
increment_x();
get_x_squared(); // 9
increment_x();
get_x_squared(); // 16
```

--- a/tests/programs/logic.vi
+++ b/tests/programs/logic.vi
@@ -1,37 +1,37 @@
 
 fn main(&io: &IO) {
-  print_bool(&io, true);
-  print_bool(&io, true && true);
-  print_bool(&io, true || true);
-  print_bool(&io, true || false);
-  print_bool(&io, false || true);
-  print_bool(&io, true => true);
-  print_bool(&io, false => true);
-  print_bool(&io, false => false);
-  print_bool(&io, !false);
-  print_bool(&io, !!true);
+  dyn fn print_bool(bool) {
+    io.println(if bool {
+      "true"
+    } else {
+      "false"
+    })
+  }
+
+  print_bool(true);
+  print_bool(true && true);
+  print_bool(true || true);
+  print_bool(true || false);
+  print_bool(false || true);
+  print_bool(true => true);
+  print_bool(false => true);
+  print_bool(false => false);
+  print_bool(!false);
+  print_bool(!!true);
   io.println("---");
-  print_bool(&io, false);
-  print_bool(&io, false || false);
-  print_bool(&io, true && false);
-  print_bool(&io, false && true);
-  print_bool(&io, false && false);
-  print_bool(&io, true => false);
-  print_bool(&io, !true);
-  print_bool(&io, !!false);
+  print_bool(false);
+  print_bool(false || false);
+  print_bool(true && false);
+  print_bool(false && true);
+  print_bool(false && false);
+  print_bool(true => false);
+  print_bool(!true);
+  print_bool(!!false);
 
   // Test short-circuiting behavior; `~_` is an eraser node,
   // so would erase `io` if it was evaluated.
   io.println("---");
-  print_bool(&io, true || ~_);
-  print_bool(&io, false => ~_);
-  print_bool(&io, false && ~_);
-}
-
-fn print_bool(&io: &IO, bool: bool) {
-  io.println(if bool {
-    "true"
-  } else {
-    "false"
-  })
+  print_bool(true || ~_);
+  print_bool(false => ~_);
+  print_bool(false && ~_);
 }

--- a/tests/programs/option_party.vi
+++ b/tests/programs/option_party.vi
@@ -2,83 +2,83 @@
 use std::option::Option::{Option, Some, None};
 
 fn main(&io: &IO) {
+  dyn fn print_option_u32(option: Option[u32]) {
+    io.println("  " ++ match option {
+      Some(val) => "Some(" ++ val.to_string() ++ ")",
+      None => "None",
+    })
+  }
+
+  dyn fn print_u32(val: u32) {
+    io.println("  " ++ val.to_string())
+  }
+
+  dyn fn print_bool(bool: bool) {
+    io.println(if bool {
+      "  true"
+    } else {
+      "  false"
+    })
+  }
+
   io.println("map:");
-  print_option_u32(&io, None.map(fn(x: u32) x + 1));
-  print_option_u32(&io, Some(0).map(fn(x: u32) x + 1));
-  print_option_u32(&io, Some(100).map(fn(x: u32) x + 1));
+  print_option_u32(None.map(fn(x: u32) x + 1));
+  print_option_u32(Some(0).map(fn(x: u32) x + 1));
+  print_option_u32(Some(100).map(fn(x: u32) x + 1));
 
   io.println("increment_option:");
   let option = None;
   increment_option(&option);
-  print_option_u32(&io, option);
+  print_option_u32(option);
   option = Some(0);
   increment_option(&option);
-  print_option_u32(&io, option);
+  print_option_u32(option);
   increment_option(&option);
-  print_option_u32(&io, option);
+  print_option_u32(option);
 
   io.println("flatten:");
-  print_option_u32(&io, None.flatten());
-  print_option_u32(&io, Some(None).flatten());
-  print_option_u32(&io, Some(Some(123)).flatten());
+  print_option_u32(None.flatten());
+  print_option_u32(Some(None).flatten());
+  print_option_u32(Some(Some(123)).flatten());
 
   io.println("add_options:");
-  print_option_u32(&io, add_options(None, None));
-  print_option_u32(&io, add_options(Some(1), None));
-  print_option_u32(&io, add_options(None, Some(2)));
-  print_option_u32(&io, add_options(Some(1), Some(2)));
+  print_option_u32(add_options(None, None));
+  print_option_u32(add_options(Some(1), None));
+  print_option_u32(add_options(None, Some(2)));
+  print_option_u32(add_options(Some(1), Some(2)));
 
   io.println("or:");
-  print_option_u32(&io, None.or(None));
-  print_option_u32(&io, Some(1).or(None));
-  print_option_u32(&io, None.or(Some(2)));
-  print_option_u32(&io, Some(1).or(Some(2)));
+  print_option_u32(None.or(None));
+  print_option_u32(Some(1).or(None));
+  print_option_u32(None.or(Some(2)));
+  print_option_u32(Some(1).or(Some(2)));
 
   io.println("unwrap_or:");
-  print_u32(&io, None.unwrap_or(0));
-  print_u32(&io, Some(1).unwrap_or(0));
+  print_u32(None.unwrap_or(0));
+  print_u32(Some(1).unwrap_or(0));
 
   io.println("is_some:");
-  print_bool(&io, None is Some(_));
-  print_bool(&io, Some(()) is Some(_));
+  print_bool(None is Some(_));
+  print_bool(Some(()) is Some(_));
 
   io.println("is_none:");
-  print_bool(&io, None is None);
-  print_bool(&io, Some(()) is None);
+  print_bool(None is None);
+  print_bool(Some(()) is None);
 
   io.println("is_some_and:");
-  print_bool(&io, None[u32] is Some(x) && x > 0);
-  print_bool(&io, Some(0) is Some(x) && x > 0);
-  print_bool(&io, Some(1) is Some(x) && x > 0);
+  print_bool(None[u32] is Some(x) && x > 0);
+  print_bool(Some(0) is Some(x) && x > 0);
+  print_bool(Some(1) is Some(x) && x > 0);
 
   io.println("is_some_implies:");
-  print_bool(&io, None[u32] is Some(x) => x > 0);
-  print_bool(&io, Some(0) is Some(x) => x > 0);
-  print_bool(&io, Some(1) is Some(x) => x > 0);
+  print_bool(None[u32] is Some(x) => x > 0);
+  print_bool(Some(0) is Some(x) => x > 0);
+  print_bool(Some(1) is Some(x) => x > 0);
 
   io.println("unwrap:");
-  print_u32(&io, Some(42).unwrap());
-  print_u32(&io, None.unwrap());
+  print_u32(Some(42).unwrap());
+  print_u32(None.unwrap());
   io.println("io will never work again :(");
-}
-
-fn print_option_u32(&io: &IO, option: Option[u32]) {
-  io.println("  " ++ match option {
-    Some(val) => "Some(" ++ val.to_string() ++ ")",
-    None => "None",
-  })
-}
-
-fn print_u32(&io: &IO, val: u32) {
-  io.println("  " ++ val.to_string())
-}
-
-fn print_bool(&io: &IO, bool: bool) {
-  io.println(if bool {
-    "  true"
-  } else {
-    "  false"
-  })
 }
 
 fn increment_option(&option: &Option[u32]) {

--- a/tests/snaps/vine/logic/compiled.iv
+++ b/tests/snaps/vine/logic/compiled.iv
@@ -2,134 +2,134 @@
 ::main { ::logic::main }
 
 ::logic::main {
-  fn(ref(n0 n43) _)
-  ::logic::print_bool = fn(ref(n0 n1) fn(1 _))
-  ::logic::print_bool = fn(ref(n1 n3) fn(n2 _))
-  1 = ?(0 ::logic::main::2 n2)
-  ::logic::print_bool = fn(ref(n3 n5) fn(n4 _))
-  1 = ?(::logic::main::8 1 n4)
-  ::logic::print_bool = fn(ref(n5 n7) fn(n6 _))
-  1 = ?(::logic::main::13 1 n6)
-  ::logic::print_bool = fn(ref(n7 n9) fn(n8 _))
-  0 = ?(::logic::main::18 1 n8)
-  ::logic::print_bool = fn(ref(n9 n11) fn(n10 _))
-  1 = ?(1 ::logic::main::22 n10)
-  ::logic::print_bool = fn(ref(n11 n13) fn(n12 _))
-  0 = ?(1 ::logic::main::27 n12)
-  ::logic::print_bool = fn(ref(n13 n15) fn(n14 _))
-  0 = ?(1 ::logic::main::32 n14)
-  ::logic::print_bool = fn(ref(n15 n17) fn(n16 _))
-  0 = ?(1 0 n16)
-  ::logic::print_bool = fn(ref(n17 n19) fn(n18 _))
-  1 = ?(0 1 n18)
-  ::std::io::IO::println = fn(ref(n19 n21) fn(tup(3 tup(tup(45 tup(45 tup(45 n20))) n20)) _))
-  ::logic::print_bool = fn(ref(n21 n22) fn(0 _))
-  ::logic::print_bool = fn(ref(n22 n24) fn(n23 _))
-  0 = ?(::logic::main::42 1 n23)
-  ::logic::print_bool = fn(ref(n24 n26) fn(n25 _))
-  1 = ?(0 ::logic::main::46 n25)
-  ::logic::print_bool = fn(ref(n26 n28) fn(n27 _))
-  0 = ?(0 ::logic::main::51 n27)
-  ::logic::print_bool = fn(ref(n28 n30) fn(n29 _))
-  0 = ?(0 ::logic::main::56 n29)
-  ::logic::print_bool = fn(ref(n30 n32) fn(n31 _))
-  1 = ?(1 ::logic::main::61 n31)
-  ::logic::print_bool = fn(ref(n32 n34) fn(n33 _))
-  1 = ?(1 0 n33)
-  ::logic::print_bool = fn(ref(n34 n36) fn(n35 _))
-  0 = ?(0 1 n35)
-  ::std::io::IO::println = fn(ref(n36 n38) fn(tup(3 tup(tup(45 tup(45 tup(45 n37))) n37)) _))
-  ::logic::print_bool = fn(ref(n38 n40) fn(n39 _))
-  1 = ?(::logic::main::71 1 n39)
-  ::logic::print_bool = fn(ref(n40 n42) fn(n41 _))
-  0 = ?(1 ::logic::main::75 n41)
-  ::logic::print_bool = fn(ref(n42 n43) fn(n44 _))
-  0 = ?(0 ::logic::main::80 n44)
+  fn(ref(n21 n44) _)
+  1 = ?(0 ::logic::main::5 n0)
+  1 = ?(::logic::main::11 1 n1)
+  1 = ?(::logic::main::16 1 n2)
+  0 = ?(::logic::main::21 1 n3)
+  1 = ?(1 ::logic::main::25 n4)
+  0 = ?(1 ::logic::main::30 n5)
+  0 = ?(1 ::logic::main::35 n6)
+  0 = ?(1 0 n7)
+  1 = ?(0 1 n8)
+  ::std::io::IO::println = fn(ref(n31 n32) fn(tup(3 tup(tup(45 tup(45 tup(45 n9))) n9)) _))
+  0 = ?(::logic::main::45 1 n10)
+  1 = ?(0 ::logic::main::49 n11)
+  0 = ?(0 ::logic::main::54 n12)
+  0 = ?(0 ::logic::main::59 n13)
+  1 = ?(1 ::logic::main::64 n14)
+  1 = ?(1 0 n15)
+  0 = ?(0 1 n16)
+  ::std::io::IO::println = fn(ref(n40 n41) fn(tup(3 tup(tup(45 tup(45 tup(45 n17))) n17)) _))
+  1 = ?(::logic::main::74 1 n18)
+  0 = ?(1 ::logic::main::78 n19)
+  0 = ?(0 ::logic::main::83 n20)
+  ::logic::main::1 = x(n21 x(n22 fn(1 _)))
+  ::logic::main::1 = x(n22 x(n23 fn(n0 _)))
+  ::logic::main::1 = x(n23 x(n24 fn(n1 _)))
+  ::logic::main::1 = x(n24 x(n25 fn(n2 _)))
+  ::logic::main::1 = x(n25 x(n26 fn(n3 _)))
+  ::logic::main::1 = x(n26 x(n27 fn(n4 _)))
+  ::logic::main::1 = x(n27 x(n28 fn(n5 _)))
+  ::logic::main::1 = x(n28 x(n29 fn(n6 _)))
+  ::logic::main::1 = x(n29 x(n30 fn(n7 _)))
+  ::logic::main::1 = x(n30 x(n31 fn(n8 _)))
+  ::logic::main::1 = x(n32 x(n33 fn(0 _)))
+  ::logic::main::1 = x(n33 x(n34 fn(n10 _)))
+  ::logic::main::1 = x(n34 x(n35 fn(n11 _)))
+  ::logic::main::1 = x(n35 x(n36 fn(n12 _)))
+  ::logic::main::1 = x(n36 x(n37 fn(n13 _)))
+  ::logic::main::1 = x(n37 x(n38 fn(n14 _)))
+  ::logic::main::1 = x(n38 x(n39 fn(n15 _)))
+  ::logic::main::1 = x(n39 x(n40 fn(n16 _)))
+  ::logic::main::1 = x(n41 x(n42 fn(n18 _)))
+  ::logic::main::1 = x(n42 x(n43 fn(n19 _)))
+  ::logic::main::1 = x(n43 x(n44 fn(n20 _)))
 }
 
-::logic::main::2 {
+::logic::main::1 {
+  x(n2 x(n3 fn(?(::logic::main::3 ::logic::main::2 n0) n1)))
+  ::std::io::IO::println = fn(ref(n2 n3) fn(n0 n1))
+}
+
+::logic::main::2 { tup(4 tup(tup(116 tup(114 tup(117 tup(101 n0)))) n0)) }
+
+::logic::main::3 { tup(5 tup(tup(102 tup(97 tup(108 tup(115 tup(101 n0))))) n0)) }
+
+::logic::main::5 {
   n0
   1 = ?(0 1 n0)
 }
 
-::logic::main::8 {
+::logic::main::11 {
   n0
   1 = ?(0 1 n0)
 }
 
-::logic::main::13 {
+::logic::main::16 {
   n0
   0 = ?(0 1 n0)
 }
 
-::logic::main::18 {
+::logic::main::21 {
   n0
   1 = ?(0 1 n0)
 }
 
-::logic::main::22 {
+::logic::main::25 {
   n0
   1 = ?(0 1 n0)
 }
 
-::logic::main::27 {
+::logic::main::30 {
   n0
   1 = ?(0 1 n0)
 }
 
-::logic::main::32 {
+::logic::main::35 {
   n0
   0 = ?(0 1 n0)
 }
 
-::logic::main::42 {
+::logic::main::45 {
   n0
   0 = ?(0 1 n0)
 }
 
-::logic::main::46 {
+::logic::main::49 {
   n0
   0 = ?(0 1 n0)
 }
 
-::logic::main::51 {
+::logic::main::54 {
   n0
   1 = ?(0 1 n0)
 }
 
-::logic::main::56 {
+::logic::main::59 {
   n0
   0 = ?(0 1 n0)
 }
 
-::logic::main::61 {
+::logic::main::64 {
   n0
   0 = ?(0 1 n0)
 }
 
-::logic::main::71 {
+::logic::main::74 {
   n0
   _ = ?(0 1 n0)
 }
 
-::logic::main::75 {
+::logic::main::78 {
   n0
   _ = ?(0 1 n0)
 }
 
-::logic::main::80 {
+::logic::main::83 {
   n0
   _ = ?(0 1 n0)
 }
-
-::logic::print_bool {
-  fn(n0 fn(?(::logic::print_bool::2 ::logic::print_bool::1 n2) n3))
-  ::std::io::IO::println = fn(n0 fn(n2 n3))
-}
-
-::logic::print_bool::1 { tup(4 tup(tup(116 tup(114 tup(117 tup(101 n0)))) n0)) }
-
-::logic::print_bool::2 { tup(5 tup(tup(102 tup(97 tup(108 tup(115 tup(101 n0))))) n0)) }
 
 ::std::io::IO::println {
   fn(ref(n0 n3) fn(n1 _))
@@ -142,7 +142,7 @@
   ::std::io::IO::print::1 = x(n0 x(n1 x(n2 n3)))
 }
 
-::std::io::IO::print::1 { x(n0 x(n1 x(dup36(?(::std::io::IO::print::3 ::std::io::IO::print::2 x(n0 x(n1 x(n3 n2)))) n3) n2))) }
+::std::io::IO::print::1 { x(n0 x(n1 x(dup35(?(::std::io::IO::print::3 ::std::io::IO::print::2 x(n0 x(n1 x(n3 n2)))) n3) n2))) }
 
 ::std::io::IO::print::2 {
   x(n1 x(n3 x(@sub(1 n4) tup(n0 n5))))

--- a/tests/snaps/vine/logic/stats.txt
+++ b/tests/snaps/vine/logic/stats.txt
@@ -1,7 +1,7 @@
 
 Interactions
-  Total                  2_950
-  Annihilate             1_568
+  Total                  2_971
+  Annihilate             1_589
   Commute                    0
   Copy                     147
   Erase                    384
@@ -10,6 +10,6 @@ Interactions
   Branch                   168
 
 Memory
-  Heap                   2_176 B
-  Allocated             58_752 B
-  Freed                 58_752 B
+  Heap                   1_904 B
+  Allocated             59_424 B
+  Freed                 59_424 B

--- a/tests/snaps/vine/option_party/compiled.iv
+++ b/tests/snaps/vine/option_party/compiled.iv
@@ -2,149 +2,149 @@
 ::main { ::option_party::main }
 
 ::option_party::main {
-  fn(ref(n0 n103) _)
-  ::std::io::IO::println = fn(ref(n0 n2) fn(tup(4 tup(tup(109 tup(97 tup(112 tup(58 n1)))) n1)) _))
-  ::option_party::print_option_u32 = fn(ref(n2 n5) fn(n4 _))
-  ::std::option::Option::map = fn(::std::option::Option::None fn(fn(@add(1 n3) n3) n4))
-  ::option_party::print_option_u32 = fn(ref(n5 n9) fn(n8 _))
-  ::std::option::Option::Some = fn(0 n6)
-  ::std::option::Option::map = fn(n6 fn(fn(@add(1 n7) n7) n8))
-  ::option_party::print_option_u32 = fn(ref(n9 n13) fn(n12 _))
-  ::std::option::Option::Some = fn(100 n10)
-  ::std::option::Option::map = fn(n10 fn(fn(@add(1 n11) n11) n12))
-  ::std::io::IO::println = fn(ref(n13 n15) fn(tup(17 tup(tup(105 tup(110 tup(99 tup(114 tup(101 tup(109 tup(101 tup(110 tup(116 tup(95 tup(111 tup(112 tup(116 tup(105 tup(111 tup(110 tup(58 n14))))))))))))))))) n14)) _))
-  ::option_party::increment_option = fn(ref(::std::option::Option::None n16) _)
-  ::option_party::print_option_u32 = fn(ref(n15 n18) fn(n16 _))
-  ::std::option::Option::Some = fn(0 n17)
-  ::option_party::increment_option = fn(ref(n17 dup5(n105 n106)) _)
-  ::option_party::print_option_u32 = fn(ref(n18 n19) fn(n105 _))
-  ::option_party::increment_option = fn(ref(n106 n20) _)
-  ::option_party::print_option_u32 = fn(ref(n19 n21) fn(n20 _))
-  ::std::io::IO::println = fn(ref(n21 n23) fn(tup(8 tup(tup(102 tup(108 tup(97 tup(116 tup(116 tup(101 tup(110 tup(58 n22)))))))) n22)) _))
-  ::option_party::print_option_u32 = fn(ref(n23 n25) fn(n24 _))
-  ::std::option::Option::flatten = fn(::std::option::Option::None n24)
-  ::option_party::print_option_u32 = fn(ref(n25 n28) fn(n27 _))
-  ::std::option::Option::Some = fn(::std::option::Option::None n26)
-  ::std::option::Option::flatten = fn(n26 n27)
-  ::option_party::print_option_u32 = fn(ref(n28 n32) fn(n31 _))
-  ::std::option::Option::Some = fn(123 n29)
-  ::std::option::Option::Some = fn(n29 n30)
-  ::std::option::Option::flatten = fn(n30 n31)
-  ::std::io::IO::println = fn(ref(n32 n34) fn(tup(12 tup(tup(97 tup(100 tup(100 tup(95 tup(111 tup(112 tup(116 tup(105 tup(111 tup(110 tup(115 tup(58 n33)))))))))))) n33)) _))
-  ::option_party::print_option_u32 = fn(ref(n34 n36) fn(n35 _))
-  ::option_party::add_options = fn(::std::option::Option::None fn(::std::option::Option::None n35))
-  ::option_party::print_option_u32 = fn(ref(n36 n39) fn(n38 _))
-  ::std::option::Option::Some = fn(1 n37)
-  ::option_party::add_options = fn(n37 fn(::std::option::Option::None n38))
-  ::option_party::print_option_u32 = fn(ref(n39 n42) fn(n41 _))
-  ::option_party::add_options = fn(::std::option::Option::None fn(n40 n41))
-  ::std::option::Option::Some = fn(2 n40)
-  ::option_party::print_option_u32 = fn(ref(n42 n46) fn(n45 _))
-  ::std::option::Option::Some = fn(1 n43)
-  ::option_party::add_options = fn(n43 fn(n44 n45))
-  ::std::option::Option::Some = fn(2 n44)
-  ::std::io::IO::println = fn(ref(n46 n48) fn(tup(3 tup(tup(111 tup(114 tup(58 n47))) n47)) _))
-  ::option_party::print_option_u32 = fn(ref(n48 n50) fn(n49 _))
-  ::std::option::Option::or = fn(::std::option::Option::None fn(::std::option::Option::None n49))
-  ::option_party::print_option_u32 = fn(ref(n50 n53) fn(n52 _))
-  ::std::option::Option::Some = fn(1 n51)
-  ::std::option::Option::or = fn(n51 fn(::std::option::Option::None n52))
-  ::option_party::print_option_u32 = fn(ref(n53 n56) fn(n55 _))
-  ::std::option::Option::or = fn(::std::option::Option::None fn(n54 n55))
-  ::std::option::Option::Some = fn(2 n54)
-  ::option_party::print_option_u32 = fn(ref(n56 n60) fn(n59 _))
-  ::std::option::Option::Some = fn(1 n57)
-  ::std::option::Option::or = fn(n57 fn(n58 n59))
-  ::std::option::Option::Some = fn(2 n58)
-  ::std::io::IO::println = fn(ref(n60 n62) fn(tup(10 tup(tup(117 tup(110 tup(119 tup(114 tup(97 tup(112 tup(95 tup(111 tup(114 tup(58 n61)))))))))) n61)) _))
-  ::option_party::print_u32 = fn(ref(n62 n64) fn(n63 _))
-  ::std::option::Option::unwrap_or = fn(::std::option::Option::None fn(0 n63))
-  ::option_party::print_u32 = fn(ref(n64 n67) fn(n66 _))
-  ::std::option::Option::Some = fn(1 n65)
-  ::std::option::Option::unwrap_or = fn(n65 fn(0 n66))
-  ::std::io::IO::println = fn(ref(n67 n69) fn(tup(8 tup(tup(105 tup(115 tup(95 tup(115 tup(111 tup(109 tup(101 tup(58 n68)))))))) n68)) _))
-  ::option_party::print_bool = fn(ref(n69 n71) fn(n70 _))
-  ::std::option::Option::None = enum(::option_party::main::1 enum(0 n70))
-  ::option_party::print_bool = fn(ref(n71 n73) fn(n72 _))
-  ::std::option::Option::Some = fn(_ enum(::option_party::main::3 enum(0 n72)))
-  ::std::io::IO::println = fn(ref(n73 n75) fn(tup(8 tup(tup(105 tup(115 tup(95 tup(110 tup(111 tup(110 tup(101 tup(58 n74)))))))) n74)) _))
-  ::option_party::print_bool = fn(ref(n75 n77) fn(n76 _))
-  ::std::option::Option::None = enum(::option_party::main::5 enum(1 n76))
-  ::option_party::print_bool = fn(ref(n77 n79) fn(n78 _))
-  ::std::option::Option::Some = fn(_ enum(::option_party::main::7 enum(1 n78)))
-  ::std::io::IO::println = fn(ref(n79 n81) fn(tup(12 tup(tup(105 tup(115 tup(95 tup(115 tup(111 tup(109 tup(101 tup(95 tup(97 tup(110 tup(100 tup(58 n80)))))))))))) n80)) _))
-  ::option_party::print_bool = fn(ref(n81 n83) fn(n82 _))
-  ::std::option::Option::None = enum(::option_party::main::10 enum(0 n82))
-  ::option_party::print_bool = fn(ref(n83 n85) fn(n84 _))
-  ::std::option::Option::Some = fn(0 enum(::option_party::main::15 enum(0 n84)))
-  ::option_party::print_bool = fn(ref(n85 n87) fn(n86 _))
-  ::std::option::Option::Some = fn(1 enum(::option_party::main::20 enum(0 n86)))
-  ::std::io::IO::println = fn(ref(n87 n89) fn(tup(16 tup(tup(105 tup(115 tup(95 tup(115 tup(111 tup(109 tup(101 tup(95 tup(105 tup(109 tup(112 tup(108 tup(105 tup(101 tup(115 tup(58 n88)))))))))))))))) n88)) _))
-  ::option_party::print_bool = fn(ref(n89 n91) fn(n90 _))
-  ::std::option::Option::None = enum(::option_party::main::25 enum(1 n90))
-  ::option_party::print_bool = fn(ref(n91 n93) fn(n92 _))
-  ::std::option::Option::Some = fn(0 enum(::option_party::main::30 enum(1 n92)))
-  ::option_party::print_bool = fn(ref(n93 n95) fn(n94 _))
-  ::std::option::Option::Some = fn(1 enum(::option_party::main::35 enum(1 n94)))
-  ::std::io::IO::println = fn(ref(n95 n97) fn(tup(7 tup(tup(117 tup(110 tup(119 tup(114 tup(97 tup(112 tup(58 n96))))))) n96)) _))
-  ::option_party::print_u32 = fn(ref(n97 n100) fn(n99 _))
-  ::std::option::Option::Some = fn(42 n98)
-  ::std::option::Option::unwrap = fn(n98 n99)
-  ::option_party::print_u32 = fn(ref(n100 n102) fn(n101 _))
-  ::std::option::Option::unwrap = fn(::std::option::Option::None n101)
-  ::std::io::IO::println = fn(ref(n102 n103) fn(tup(27 tup(tup(105 tup(111 tup(32 tup(119 tup(105 tup(108 tup(108 tup(32 tup(110 tup(101 tup(118 tup(101 tup(114 tup(32 tup(119 tup(111 tup(114 tup(107 tup(32 tup(97 tup(103 tup(97 tup(105 tup(110 tup(32 tup(58 tup(40 n104))))))))))))))))))))))))))) n104)) _))
+  fn(ref(n0 n61) _)
+  ::std::io::IO::println = fn(ref(n0 n63) fn(tup(4 tup(tup(109 tup(97 tup(112 tup(58 n1)))) n1)) _))
+  ::std::option::Option::map = fn(::std::option::Option::None fn(fn(@add(1 n2) n2) n3))
+  ::std::option::Option::Some = fn(0 n4)
+  ::std::option::Option::map = fn(n4 fn(fn(@add(1 n5) n5) n6))
+  ::std::option::Option::Some = fn(100 n7)
+  ::std::option::Option::map = fn(n7 fn(fn(@add(1 n8) n8) n9))
+  ::std::io::IO::println = fn(ref(n66 n67) fn(tup(17 tup(tup(105 tup(110 tup(99 tup(114 tup(101 tup(109 tup(101 tup(110 tup(116 tup(95 tup(111 tup(112 tup(116 tup(105 tup(111 tup(110 tup(58 n10))))))))))))))))) n10)) _))
+  ::option_party::increment_option = fn(ref(::std::option::Option::None n11) _)
+  ::std::option::Option::Some = fn(0 n12)
+  ::option_party::increment_option = fn(ref(n12 dup9(n69 n70)) _)
+  ::option_party::increment_option = fn(ref(n70 n13) _)
+  ::std::io::IO::println = fn(ref(n72 n73) fn(tup(8 tup(tup(102 tup(108 tup(97 tup(116 tup(116 tup(101 tup(110 tup(58 n14)))))))) n14)) _))
+  ::std::option::Option::flatten = fn(::std::option::Option::None n15)
+  ::std::option::Option::Some = fn(::std::option::Option::None n16)
+  ::std::option::Option::flatten = fn(n16 n17)
+  ::std::option::Option::Some = fn(123 n18)
+  ::std::option::Option::Some = fn(n18 n19)
+  ::std::option::Option::flatten = fn(n19 n20)
+  ::std::io::IO::println = fn(ref(n76 n77) fn(tup(12 tup(tup(97 tup(100 tup(100 tup(95 tup(111 tup(112 tup(116 tup(105 tup(111 tup(110 tup(115 tup(58 n21)))))))))))) n21)) _))
+  ::option_party::add_options = fn(::std::option::Option::None fn(::std::option::Option::None n22))
+  ::std::option::Option::Some = fn(1 n23)
+  ::option_party::add_options = fn(n23 fn(::std::option::Option::None n24))
+  ::option_party::add_options = fn(::std::option::Option::None fn(n25 n26))
+  ::std::option::Option::Some = fn(2 n25)
+  ::std::option::Option::Some = fn(1 n27)
+  ::option_party::add_options = fn(n27 fn(n28 n29))
+  ::std::option::Option::Some = fn(2 n28)
+  ::std::io::IO::println = fn(ref(n81 n82) fn(tup(3 tup(tup(111 tup(114 tup(58 n30))) n30)) _))
+  ::std::option::Option::or = fn(::std::option::Option::None fn(::std::option::Option::None n31))
+  ::std::option::Option::Some = fn(1 n32)
+  ::std::option::Option::or = fn(n32 fn(::std::option::Option::None n33))
+  ::std::option::Option::or = fn(::std::option::Option::None fn(n34 n35))
+  ::std::option::Option::Some = fn(2 n34)
+  ::std::option::Option::Some = fn(1 n36)
+  ::std::option::Option::or = fn(n36 fn(n37 n38))
+  ::std::option::Option::Some = fn(2 n37)
+  ::std::io::IO::println = fn(ref(n86 n87) fn(tup(10 tup(tup(117 tup(110 tup(119 tup(114 tup(97 tup(112 tup(95 tup(111 tup(114 tup(58 n39)))))))))) n39)) _))
+  ::std::option::Option::unwrap_or = fn(::std::option::Option::None fn(0 n40))
+  ::std::option::Option::Some = fn(1 n41)
+  ::std::option::Option::unwrap_or = fn(n41 fn(0 n42))
+  ::std::io::IO::println = fn(ref(n89 n90) fn(tup(8 tup(tup(105 tup(115 tup(95 tup(115 tup(111 tup(109 tup(101 tup(58 n43)))))))) n43)) _))
+  ::std::option::Option::None = enum(::option_party::main::8 enum(0 n44))
+  ::std::option::Option::Some = fn(_ enum(::option_party::main::10 enum(0 n45)))
+  ::std::io::IO::println = fn(ref(n92 n93) fn(tup(8 tup(tup(105 tup(115 tup(95 tup(110 tup(111 tup(110 tup(101 tup(58 n46)))))))) n46)) _))
+  ::std::option::Option::None = enum(::option_party::main::12 enum(1 n47))
+  ::std::option::Option::Some = fn(_ enum(::option_party::main::14 enum(1 n48)))
+  ::std::io::IO::println = fn(ref(n95 n96) fn(tup(12 tup(tup(105 tup(115 tup(95 tup(115 tup(111 tup(109 tup(101 tup(95 tup(97 tup(110 tup(100 tup(58 n49)))))))))))) n49)) _))
+  ::std::option::Option::None = enum(::option_party::main::17 enum(0 n50))
+  ::std::option::Option::Some = fn(0 enum(::option_party::main::22 enum(0 n51)))
+  ::std::option::Option::Some = fn(1 enum(::option_party::main::27 enum(0 n52)))
+  ::std::io::IO::println = fn(ref(n99 n100) fn(tup(16 tup(tup(105 tup(115 tup(95 tup(115 tup(111 tup(109 tup(101 tup(95 tup(105 tup(109 tup(112 tup(108 tup(105 tup(101 tup(115 tup(58 n53)))))))))))))))) n53)) _))
+  ::std::option::Option::None = enum(::option_party::main::32 enum(1 n54))
+  ::std::option::Option::Some = fn(0 enum(::option_party::main::37 enum(1 n55)))
+  ::std::option::Option::Some = fn(1 enum(::option_party::main::42 enum(1 n56)))
+  ::std::io::IO::println = fn(ref(n103 n104) fn(tup(7 tup(tup(117 tup(110 tup(119 tup(114 tup(97 tup(112 tup(58 n57))))))) n57)) _))
+  ::std::option::Option::Some = fn(42 n58)
+  ::std::option::Option::unwrap = fn(n58 n59)
+  ::std::option::Option::unwrap = fn(::std::option::Option::None n60)
+  ::std::io::IO::println = fn(ref(n106 n61) fn(tup(27 tup(tup(105 tup(111 tup(32 tup(119 tup(105 tup(108 tup(108 tup(32 tup(110 tup(101 tup(118 tup(101 tup(114 tup(32 tup(119 tup(111 tup(114 tup(107 tup(32 tup(97 tup(103 tup(97 tup(105 tup(110 tup(32 tup(58 tup(40 n62))))))))))))))))))))))))))) n62)) _))
+  ::option_party::main::1 = x(n63 x(n64 fn(n3 _)))
+  ::option_party::main::1 = x(n64 x(n65 fn(n6 _)))
+  ::option_party::main::1 = x(n65 x(n66 fn(n9 _)))
+  ::option_party::main::1 = x(n67 x(n68 fn(n11 _)))
+  ::option_party::main::1 = x(n68 x(n71 fn(n69 _)))
+  ::option_party::main::1 = x(n71 x(n72 fn(n13 _)))
+  ::option_party::main::1 = x(n73 x(n74 fn(n15 _)))
+  ::option_party::main::1 = x(n74 x(n75 fn(n17 _)))
+  ::option_party::main::1 = x(n75 x(n76 fn(n20 _)))
+  ::option_party::main::1 = x(n77 x(n78 fn(n22 _)))
+  ::option_party::main::1 = x(n78 x(n79 fn(n24 _)))
+  ::option_party::main::1 = x(n79 x(n80 fn(n26 _)))
+  ::option_party::main::1 = x(n80 x(n81 fn(n29 _)))
+  ::option_party::main::1 = x(n82 x(n83 fn(n31 _)))
+  ::option_party::main::1 = x(n83 x(n84 fn(n33 _)))
+  ::option_party::main::1 = x(n84 x(n85 fn(n35 _)))
+  ::option_party::main::1 = x(n85 x(n86 fn(n38 _)))
+  ::option_party::main::4 = x(n87 x(n88 fn(n40 _)))
+  ::option_party::main::4 = x(n88 x(n89 fn(n42 _)))
+  ::option_party::main::5 = x(n90 x(n91 fn(n44 _)))
+  ::option_party::main::5 = x(n91 x(n92 fn(n45 _)))
+  ::option_party::main::5 = x(n93 x(n94 fn(n47 _)))
+  ::option_party::main::5 = x(n94 x(n95 fn(n48 _)))
+  ::option_party::main::5 = x(n96 x(n97 fn(n50 _)))
+  ::option_party::main::5 = x(n97 x(n98 fn(n51 _)))
+  ::option_party::main::5 = x(n98 x(n99 fn(n52 _)))
+  ::option_party::main::5 = x(n100 x(n101 fn(n54 _)))
+  ::option_party::main::5 = x(n101 x(n102 fn(n55 _)))
+  ::option_party::main::5 = x(n102 x(n103 fn(n56 _)))
+  ::option_party::main::4 = x(n104 x(n105 fn(n59 _)))
+  ::option_party::main::4 = x(n105 x(n106 fn(n60 _)))
 }
 
-::option_party::main::1 { enum(_ 1) }
-
-::option_party::main::3 { enum(_ 1) }
-
-::option_party::main::5 { enum(_ 0) }
-
-::option_party::main::7 { enum(_ 0) }
-
-::option_party::main::10 { enum(@lt$(0 ?(0 1 n0)) n0) }
-
-::option_party::main::15 { enum(@lt$(0 ?(0 1 n0)) n0) }
-
-::option_party::main::20 { enum(@lt$(0 ?(0 1 n0)) n0) }
-
-::option_party::main::25 { enum(@lt$(0 ?(0 1 n0)) n0) }
-
-::option_party::main::30 { enum(@lt$(0 ?(0 1 n0)) n0) }
-
-::option_party::main::35 { enum(@lt$(0 ?(0 1 n0)) n0) }
-
-::option_party::print_option_u32 {
-  fn(n0 fn(enum(::option_party::print_option_u32::1 enum(::option_party::print_option_u32::2 n3)) n5))
-  ::std::io::IO::println = fn(n0 fn(n4 n5))
-  ::std::list::List::concat = fn(tup(2 tup(tup(32 tup(32 n2)) n2)) fn(n3 n4))
+::option_party::main::1 {
+  x(n4 x(n5 fn(enum(::option_party::main::2 enum(::option_party::main::3 n1)) n3)))
+  ::std::io::IO::println = fn(ref(n4 n5) fn(n2 n3))
+  ::std::list::List::concat = fn(tup(2 tup(tup(32 tup(32 n0)) n0)) fn(n1 n2))
 }
 
-::option_party::print_option_u32::1 {
+::option_party::main::2 {
   enum(n1 n5)
   ::std::u32::to_string = fn(n1 n2)
   ::std::list::List::concat = fn(tup(5 tup(tup(83 tup(111 tup(109 tup(101 tup(40 n0))))) n0)) fn(n2 n4))
   ::std::list::List::concat = fn(n4 fn(tup(1 tup(tup(41 n3) n3)) n5))
 }
 
-::option_party::print_option_u32::2 { tup(4 tup(tup(78 tup(111 tup(110 tup(101 n0)))) n0)) }
+::option_party::main::3 { tup(4 tup(tup(78 tup(111 tup(110 tup(101 n0)))) n0)) }
 
-::option_party::print_u32 {
-  fn(n0 fn(n3 n6))
-  ::std::io::IO::println = fn(n0 fn(n5 n6))
-  ::std::u32::to_string = fn(n3 n4)
-  ::std::list::List::concat = fn(tup(2 tup(tup(32 tup(32 n2)) n2)) fn(n4 n5))
+::option_party::main::4 {
+  x(n5 x(n6 fn(n1 n4)))
+  ::std::io::IO::println = fn(ref(n5 n6) fn(n3 n4))
+  ::std::u32::to_string = fn(n1 n2)
+  ::std::list::List::concat = fn(tup(2 tup(tup(32 tup(32 n0)) n0)) fn(n2 n3))
 }
 
-::option_party::print_bool {
-  fn(n0 fn(?(::option_party::print_bool::2 ::option_party::print_bool::1 n2) n3))
-  ::std::io::IO::println = fn(n0 fn(n2 n3))
+::option_party::main::5 {
+  x(n2 x(n3 fn(?(::option_party::main::7 ::option_party::main::6 n0) n1)))
+  ::std::io::IO::println = fn(ref(n2 n3) fn(n0 n1))
 }
 
-::option_party::print_bool::1 { tup(6 tup(tup(32 tup(32 tup(116 tup(114 tup(117 tup(101 n0)))))) n0)) }
+::option_party::main::6 { tup(6 tup(tup(32 tup(32 tup(116 tup(114 tup(117 tup(101 n0)))))) n0)) }
 
-::option_party::print_bool::2 { tup(7 tup(tup(32 tup(32 tup(102 tup(97 tup(108 tup(115 tup(101 n0))))))) n0)) }
+::option_party::main::7 { tup(7 tup(tup(32 tup(32 tup(102 tup(97 tup(108 tup(115 tup(101 n0))))))) n0)) }
+
+::option_party::main::8 { enum(_ 1) }
+
+::option_party::main::10 { enum(_ 1) }
+
+::option_party::main::12 { enum(_ 0) }
+
+::option_party::main::14 { enum(_ 0) }
+
+::option_party::main::17 { enum(@lt$(0 ?(0 1 n0)) n0) }
+
+::option_party::main::22 { enum(@lt$(0 ?(0 1 n0)) n0) }
+
+::option_party::main::27 { enum(@lt$(0 ?(0 1 n0)) n0) }
+
+::option_party::main::32 { enum(@lt$(0 ?(0 1 n0)) n0) }
+
+::option_party::main::37 { enum(@lt$(0 ?(0 1 n0)) n0) }
+
+::option_party::main::42 { enum(@lt$(0 ?(0 1 n0)) n0) }
 
 ::option_party::increment_option {
   fn(n0 _)
@@ -170,7 +170,7 @@
   ::std::io::IO::print::1 = x(n0 x(n1 x(n2 n3)))
 }
 
-::std::io::IO::print::1 { x(n0 x(n1 x(dup88(?(::std::io::IO::print::3 ::std::io::IO::print::2 x(n0 x(n1 x(n3 n2)))) n3) n2))) }
+::std::io::IO::print::1 { x(n0 x(n1 x(dup85(?(::std::io::IO::print::3 ::std::io::IO::print::2 x(n0 x(n1 x(n3 n2)))) n3) n2))) }
 
 ::std::io::IO::print::2 {
   x(n1 x(n3 x(@sub(1 n4) tup(n0 n5))))
@@ -242,17 +242,17 @@
 
 ::std::option::Option::unwrap::1 { enum(n0 n0) }
 
-::std::u32::to_string { fn(dup315(?(::std::u32::to_string::5 ::std::u32::to_string::1 x(n1 n0)) n1) n0) }
+::std::u32::to_string { fn(dup312(?(::std::u32::to_string::5 ::std::u32::to_string::1 x(n1 n0)) n1) n0) }
 
 ::std::u32::to_string::1 {
   x(n1 n2)
   ::std::u32::to_string::2 = x(n1 x(tup(0 tup(n0 n0)) n2))
 }
 
-::std::u32::to_string::2 { x(dup315(?(::std::u32::to_string::4 ::std::u32::to_string::3 x(n2 n0)) n2) n0) }
+::std::u32::to_string::2 { x(dup312(?(::std::u32::to_string::4 ::std::u32::to_string::3 x(n2 n0)) n2) n0) }
 
 ::std::u32::to_string::3 {
-  x(dup315(@rem(10 n0) @div(10 n4)) x(n3 n6))
+  x(dup312(@rem(10 n0) @div(10 n4)) x(n3 n6))
   48 = @add(n0 n1)
   ::std::list::List::concat = fn(tup(1 tup(tup(n1 n2) n2)) fn(n3 n5))
   ::std::u32::to_string::2 = x(n4 x(n5 n6))

--- a/tests/snaps/vine/option_party/stats.txt
+++ b/tests/snaps/vine/option_party/stats.txt
@@ -1,7 +1,7 @@
 
 Interactions
-  Total                 10_004
-  Annihilate             5_472
+  Total                 10_035
+  Annihilate             5_503
   Commute                    3
   Copy                     542
   Erase                  1_130
@@ -10,6 +10,6 @@ Interactions
   Branch                   435
 
 Memory
-  Heap                   6_368 B
-  Allocated            204_048 B
-  Freed                204_048 B
+  Heap                   7_216 B
+  Allocated            204_880 B
+  Freed                204_880 B

--- a/vine/src/ast.rs
+++ b/vine/src/ast.rs
@@ -159,6 +159,7 @@ pub struct Stmt {
 #[derive(Debug, Clone)]
 pub enum StmtKind {
   Let(LetStmt),
+  DynFn(DynFnStmt),
   Expr(Expr, bool),
   Item(Item),
   Empty,
@@ -169,6 +170,15 @@ pub struct LetStmt {
   pub bind: Pat,
   pub ty: Option<Ty>,
   pub init: Option<Expr>,
+}
+
+#[derive(Debug, Clone)]
+pub struct DynFnStmt {
+  pub name: Ident,
+  pub dyn_id: Option<usize>,
+  pub params: Vec<(Pat, Option<Ty>)>,
+  pub ret: Option<Ty>,
+  pub body: Block,
 }
 
 #[derive(Default, Clone)]
@@ -188,6 +198,8 @@ pub enum ExprKind {
   Path(GenericPath),
   #[class(place)]
   Local(usize),
+  #[class(value)]
+  DynFn(usize),
   #[class(value)]
   Block(Block),
   #[class(value)]

--- a/vine/src/checker.rs
+++ b/vine/src/checker.rs
@@ -33,6 +33,7 @@ pub struct Checker<'r> {
 pub(crate) struct CheckerState {
   pub(crate) vars: Vec<Result<Type, Span>>,
   pub(crate) locals: HashMap<usize, Var>,
+  pub(crate) dyn_fns: HashMap<usize, Type>,
 }
 
 impl<'r> Checker<'r> {
@@ -119,6 +120,21 @@ impl<'r> Checker<'r> {
             self.check_expr_form_type(value, Form::Value, &mut ty);
           }
         }
+        StmtKind::DynFn(d) => {
+          let params = d
+            .params
+            .iter_mut()
+            .map(|(p, a)| self.check_pat_annotation(p, a.as_mut(), Form::Value, false))
+            .collect();
+          let mut ret = d
+            .ret
+            .as_mut()
+            .map(|t| self.hydrate_type(t, true))
+            .unwrap_or_else(|| self.new_var(d.body.span));
+          self.check_block_type(&mut d.body, &mut ret);
+          self.state.dyn_fns.insert(d.dyn_id.unwrap(), Type::Fn(params, Box::new(ret)));
+          ty = Type::UNIT;
+        }
         StmtKind::Expr(e, semi) => {
           ty = self.check_expr_form(e, Form::Value);
           if *semi {
@@ -129,6 +145,17 @@ impl<'r> Checker<'r> {
       }
     }
     ty
+  }
+
+  fn check_block_type(&mut self, block: &mut Block, ty: &mut Type) {
+    let mut found = self.check_block(block);
+    if !self.unify(&mut found, ty) {
+      self.diags.add(Diag::ExpectedTypeFound {
+        span: block.span,
+        expected: self.display_type(ty),
+        found: self.display_type(&found),
+      });
+    }
   }
 
   fn hydrate_type(&mut self, ty: &mut Ty, inference: bool) -> Type {

--- a/vine/src/checker/check_expr.rs
+++ b/vine/src/checker/check_expr.rs
@@ -79,6 +79,7 @@ impl Checker<'_> {
     let span = expr.span;
     match &mut expr.kind {
       ExprKind![error || place || space || synthetic] => unreachable!(),
+      ExprKind::DynFn(x) => self.state.dyn_fns[x].clone(),
       ExprKind::Path(path) => report!(self.diags, expr.kind; self.typeof_value_def(path)),
       ExprKind::Block(block) => self.check_block(block),
       ExprKind::Assign(space, value) => {

--- a/vine/src/compile.rs
+++ b/vine/src/compile.rs
@@ -1,4 +1,4 @@
-use std::collections::{BTreeMap, BTreeSet, VecDeque};
+use std::collections::{BTreeMap, BTreeSet, HashMap, VecDeque};
 
 use bitflags::bitflags;
 
@@ -55,6 +55,8 @@ pub struct Compiler<'d> {
   return_target: Option<(Local, ForkId)>,
   loop_target: Option<(Local, ForkId, StageId)>,
 
+  dyn_fns: HashMap<usize, (Local, StageId)>,
+
   dup_labels: usize,
 
   concat: Option<String>,
@@ -78,6 +80,7 @@ impl<'d> Compiler<'d> {
       net: Default::default(),
       return_target: Default::default(),
       loop_target: Default::default(),
+      dyn_fns: Default::default(),
       dup_labels: Default::default(),
       concat,
     }

--- a/vine/src/compile/build_stages.rs
+++ b/vine/src/compile/build_stages.rs
@@ -51,6 +51,11 @@ impl Compiler<'_> {
     match &expr.kind {
       ExprKind![sugar || error || !value] => unreachable!("{expr:?}"),
 
+      ExprKind::DynFn(d) => {
+        let (local, stage) = self.dyn_fns[d];
+        self.goto(stage);
+        self.move_local(local)
+      }
       ExprKind::Paren(e) => self.lower_expr_value(e),
       ExprKind::U32(num) => Port::U32(*num),
       ExprKind::F32(num) => Port::F32(*num),
@@ -308,6 +313,19 @@ impl Compiler<'_> {
         let i = l.init.as_ref().map(|x| self.lower_expr_value(x)).unwrap_or(Port::Erase);
         let b = self.lower_pat_value(&l.bind);
         self.net.link(b, i);
+        Port::Erase
+      }
+      StmtKind::DynFn(d) => {
+        let f = self.new_local();
+        self.new_fork(|self_| {
+          let i = self_.new_interface();
+          let s = self_.new_stage(i, |self_, _| {
+            let p = self_._lower_fn(&d.params, |self_| self_.lower_block(&d.body));
+            self_.set_local_to(f, p);
+            true
+          });
+          self_.dyn_fns.insert(d.dyn_id.unwrap(), (f, s));
+        });
         Port::Erase
       }
       StmtKind::Expr(t, semi) => {

--- a/vine/src/compile/build_stages/control_flow.rs
+++ b/vine/src/compile/build_stages/control_flow.rs
@@ -5,13 +5,21 @@ use crate::{
 
 impl Compiler<'_> {
   pub(super) fn lower_fn(&mut self, params: &[(Pat, Option<Ty>)], body: &Expr) -> Port {
+    self._lower_fn(params, |self_| self_.lower_expr_value(body))
+  }
+
+  pub(super) fn _lower_fn(
+    &mut self,
+    params: &[(Pat, Option<Ty>)],
+    body: impl FnOnce(&mut Self) -> Port,
+  ) -> Port {
     let func = self.net.new_wire();
     let res = self.apply_combs("fn", func.0, params.iter().map(|x| &x.0), Self::lower_pat_value);
     let result = self.new_local();
     let orig = self.cur_id;
     let old_loop = self.loop_target.take();
     let old_return = self.return_target.replace((result, self.cur_fork()));
-    let body = self.lower_expr_value(body);
+    let body = body(self);
     self.return_target = old_return;
     self.loop_target = old_loop;
     self.set_local_to(result, body);

--- a/vine/src/fmt.rs
+++ b/vine/src/fmt.rs
@@ -209,6 +209,14 @@ impl<'src> Formatter<'src> {
         },
         Doc(";"),
       ]),
+      StmtKind::DynFn(d) => Doc::concat([
+        Doc("dyn fn "),
+        Doc(d.name),
+        self.fmt_params(&d.params),
+        self.fmt_return_ty(d.ret.as_ref()),
+        Doc(" "),
+        self.fmt_block(&d.body, true),
+      ]),
       StmtKind::Expr(expr, false) => self.fmt_expr(expr),
       StmtKind::Expr(expr, true) => Doc::concat([self.fmt_expr(expr), Doc(";")]),
       StmtKind::Item(item) => self.fmt_item(item),
@@ -238,7 +246,7 @@ impl<'src> Formatter<'src> {
 
   fn fmt_expr(&self, expr: &Expr) -> Doc<'src> {
     match &expr.kind {
-      ExprKind![synthetic || error] | ExprKind::Local(_) => unreachable!(),
+      ExprKind![synthetic || error] | ExprKind::Local(_) | ExprKind::DynFn(_) => unreachable!(),
       ExprKind::Paren(p) => Doc::paren(self.fmt_expr(p)),
       ExprKind::Hole => Doc("_"),
       ExprKind::Path(path) => self.fmt_generic_path(path),

--- a/vine/src/lexer.rs
+++ b/vine/src/lexer.rs
@@ -114,6 +114,8 @@ pub enum Token {
   Move,
   #[token("let")]
   Let,
+  #[token("dyn")]
+  Dyn,
   #[token("const")]
   Const,
   #[token("defer")]

--- a/vine/src/parser.rs
+++ b/vine/src/parser.rs
@@ -8,10 +8,10 @@ use vine_util::{
 
 use crate::{
   ast::{
-    Attr, AttrKind, BinaryOp, Block, Builtin, ComparisonOp, ConstItem, Enum, Expr, ExprKind,
-    FnItem, GenericPath, Ident, InlineIvy, Item, ItemKind, LetStmt, LogicalOp, ModItem, ModKind,
-    Pat, PatKind, Path, PatternItem, Span, Stmt, StmtKind, StructItem, Ty, TyKind, TypeItem,
-    UseItem, UseTree, Variant,
+    Attr, AttrKind, BinaryOp, Block, Builtin, ComparisonOp, ConstItem, DynFnStmt, Enum, Expr,
+    ExprKind, FnItem, GenericPath, Ident, InlineIvy, Item, ItemKind, LetStmt, LogicalOp, ModItem,
+    ModKind, Pat, PatKind, Path, PatternItem, Span, Stmt, StmtKind, StructItem, Ty, TyKind,
+    TypeItem, UseItem, UseTree, Variant,
   },
   diag::Diag,
   lexer::Token,
@@ -628,6 +628,8 @@ impl<'ctx, 'src> VineParser<'ctx, 'src> {
     let span = self.start_span();
     let kind = if self.check(Token::Let) {
       StmtKind::Let(self.parse_let_stmt()?)
+    } else if self.check(Token::Dyn) {
+      StmtKind::DynFn(self.parse_dyn_fn_stmt()?)
     } else if self.eat(Token::Semi)? {
       StmtKind::Empty
     } else if let Some(item) = self.maybe_parse_item()? {
@@ -648,6 +650,16 @@ impl<'ctx, 'src> VineParser<'ctx, 'src> {
     let init = self.eat(Token::Eq)?.then(|| self.parse_expr()).transpose()?;
     self.eat(Token::Semi)?;
     Ok(LetStmt { bind, ty, init })
+  }
+
+  fn parse_dyn_fn_stmt(&mut self) -> Parse<'src, DynFnStmt> {
+    self.expect(Token::Dyn)?;
+    self.expect(Token::Fn)?;
+    let name = self.parse_ident()?;
+    let params = self.parse_delimited(PAREN_COMMA, Self::parse_pat_type)?;
+    let ret = self.eat(Token::ThinArrow)?.then(|| self.parse_type()).transpose()?;
+    let body = self.parse_block()?;
+    Ok(DynFnStmt { name, dyn_id: None, params, ret, body })
   }
 
   fn parse_path(&mut self) -> Parse<'src, Path> {

--- a/vine/src/repl.rs
+++ b/vine/src/repl.rs
@@ -83,7 +83,11 @@ impl<'ctx, 'ivm> Repl<'ctx, 'ivm> {
     let vars = HashMap::from([(io, Var { local: 0, value: Port::new_ext_val(ExtVal::IO) })]);
     let locals = BTreeMap::from([(0, io)]);
 
-    let checker_state = CheckerState { vars: vec![Ok(Type::IO)], locals: HashMap::from([(0, 0)]) };
+    let checker_state = CheckerState {
+      vars: vec![Ok(Type::IO)],
+      locals: HashMap::from([(0, 0)]),
+      dyn_fns: HashMap::new(),
+    };
 
     Ok(Repl {
       host,

--- a/vine/src/resolve/resolve_defs.rs
+++ b/vine/src/resolve/resolve_defs.rs
@@ -5,7 +5,7 @@ use std::{
 };
 
 use crate::{
-  ast::{Expr, ExprKind, GenericPath, Ident, LogicalOp, Pat, PatKind, Ty, TyKind},
+  ast::{Expr, ExprKind, GenericPath, Ident, LogicalOp, Pat, PatKind, Stmt, StmtKind, Ty, TyKind},
   diag::Diag,
   visit::{VisitMut, Visitee},
 };
@@ -25,6 +25,7 @@ impl Resolver {
       scope: HashMap::new(),
       scope_depth: 0,
       next_local: 0,
+      next_dyn_fn: 0,
     };
 
     for def_id in range {
@@ -85,19 +86,32 @@ impl Resolver {
       resolver: self,
       def,
       generics: Vec::new(),
-      scope: initial.iter().map(|(&l, &i)| (i, vec![ScopeEntry { depth: 0, local: l }])).collect(),
+      scope: initial
+        .iter()
+        .map(|(&l, &i)| (i, vec![ScopeEntry { depth: 0, binding: Binding::Local(l) }]))
+        .collect(),
       scope_depth: 0,
       next_local: *local_count,
+      next_dyn_fn: 0,
     };
     visitor.visit(visitee);
     *local_count = visitor.next_local;
     visitor.scope.into_iter().filter_map(|(k, e)| {
       let entry = e.first()?;
-      (entry.depth == 0).then_some((k, entry.local))
+      if entry.depth == 0 {
+        if let Binding::Local(l) = entry.binding {
+          Some((k, l))
+        } else {
+          None
+        }
+      } else {
+        None
+      }
     })
   }
 }
 
+#[derive(Debug)]
 struct ResolveVisitor<'a> {
   resolver: &'a mut Resolver,
   def: DefId,
@@ -105,11 +119,19 @@ struct ResolveVisitor<'a> {
   scope: HashMap<Ident, Vec<ScopeEntry>>,
   scope_depth: usize,
   next_local: usize,
+  next_dyn_fn: usize,
 }
 
+#[derive(Debug)]
 struct ScopeEntry {
   depth: usize,
-  local: usize,
+  binding: Binding,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum Binding {
+  Local(usize),
+  DynFn(usize),
 }
 
 impl VisitMut<'_> for ResolveVisitor<'_> {
@@ -123,6 +145,16 @@ impl VisitMut<'_> for ResolveVisitor<'_> {
       if stack.last().is_some_and(|x| x.depth > self.scope_depth) {
         stack.pop();
       }
+    }
+  }
+
+  fn visit_stmt(&mut self, stmt: &mut Stmt) {
+    self._visit_stmt(stmt);
+    if let StmtKind::DynFn(d) = &mut stmt.kind {
+      let id = self.next_dyn_fn;
+      self.next_dyn_fn += 1;
+      self.bind(d.name, Binding::DynFn(id));
+      d.dyn_id = Some(id);
     }
   }
 
@@ -148,13 +180,8 @@ impl VisitMut<'_> for ResolveVisitor<'_> {
       };
       let local = self.next_local;
       self.next_local += 1;
-      let stack = self.scope.entry(ident).or_default();
-      let top = stack.last_mut();
-      if top.as_ref().is_some_and(|x| x.depth == self.scope_depth) {
-        top.unwrap().local = local;
-      } else {
-        stack.push(ScopeEntry { depth: self.scope_depth, local })
-      }
+      let binding = Binding::Local(local);
+      self.bind(ident, binding);
       pat.kind = PatKind::Local(local);
     }
     self._visit_pat(pat);
@@ -185,7 +212,7 @@ impl VisitMut<'_> for ResolveVisitor<'_> {
           ExprKind::Path(path) => {
             if let Some(ident) = path.path.as_ident() {
               if let Some(bind) = self.scope.get(&ident).and_then(|x| x.last()) {
-                expr.kind = ExprKind::Local(bind.local);
+                expr.kind = bind.binding.into();
                 return;
               }
             }
@@ -222,6 +249,16 @@ impl VisitMut<'_> for ResolveVisitor<'_> {
 }
 
 impl<'a> ResolveVisitor<'a> {
+  fn bind(&mut self, ident: Ident, binding: Binding) {
+    let stack = self.scope.entry(ident).or_default();
+    let top = stack.last_mut();
+    if top.as_ref().is_some_and(|x| x.depth == self.scope_depth) {
+      top.unwrap().binding = binding;
+    } else {
+      stack.push(ScopeEntry { depth: self.scope_depth, binding })
+    }
+  }
+
   fn visit_path(&mut self, path: &mut GenericPath) -> Result<(), Diag> {
     let resolved = self.resolver.resolve_path(path.span, self.def, &path.path)?;
     path.path = self.resolver.defs[resolved].canonical.clone();
@@ -259,6 +296,15 @@ impl<'a> ResolveVisitor<'a> {
         self.visit_cond(b);
         self.exit_scope();
       }
+    }
+  }
+}
+
+impl From<Binding> for ExprKind {
+  fn from(val: Binding) -> Self {
+    match val {
+      Binding::Local(l) => ExprKind::Local(l),
+      Binding::DynFn(d) => ExprKind::DynFn(d),
     }
   }
 }

--- a/vine/src/visit.rs
+++ b/vine/src/visit.rs
@@ -54,6 +54,7 @@ pub trait VisitMut<'a> {
     match &mut expr.kind {
       ExprKind::Hole
       | ExprKind::Local(_)
+      | ExprKind::DynFn(_)
       | ExprKind::U32(_)
       | ExprKind::F32(_)
       | ExprKind::String(_)
@@ -213,6 +214,20 @@ pub trait VisitMut<'a> {
           self.visit_expr(init);
         }
         self.visit_pat(&mut l.bind);
+      }
+      StmtKind::DynFn(d) => {
+        self.enter_scope();
+        for (p, t) in &mut d.params {
+          self.visit_pat(p);
+          if let Some(t) = t {
+            self.visit_type(t);
+          }
+        }
+        if let Some(t) = &mut d.ret {
+          self.visit_type(t);
+        }
+        self.visit_block(&mut d.body);
+        self.exit_scope();
       }
       StmtKind::Expr(t, _) => self.visit_expr(t),
       StmtKind::Item(i) => self.visit_item(i),

